pkg/nlurules/listener.go

```diff
@@ -1,84 +1,108 @@
+)
 
-	current_intent string
-	current_entity string
-	current_intent_bio string
+type Utterance struct {
+	Parsed           []map[string]string
+	CurrentIntent    string
+	CurrentEntity    string
+	CurrentIntentBio string
 }
 
-func NewUtterance() Utterance {
+const (
+	// TODO: better names
+	IntentBioO = "o"
+	IntentBioB = "b"
+	IntentBioI = "i"
 
-	var l Utterance
-	l.current_intent = ""
-	l.current_entity = ""
-	l.current_intent_bio = "o"
-	return l
+	TextEmptySpace = " "
+)
 
+func NewUtterance() Utterance {
+	return Utterance{
+		CurrentIntentBio: IntentBioO,
+	}
 }
 
 type NluRuleListener struct {
 	parser.BaseAnnotationGrammarListener
 	ParsedRules []Utterance
 	utterance   Utterance
+	debug       bool
 	ch          chan Utterance
 }
 
-func (l *NluRuleListener) ExitUtterance(c *parser.UtteranceContext) {
+func NewNluRuleListener(bufSize uint64, debug bool) *NluRuleListener {
+	return &NluRuleListener{
+		ch: make(chan Utterance, bufSize),
+	}
+}
 
+func (l *NluRuleListener) ExitUtterance(c *parser.UtteranceContext) {
+	// TODO: make listener a per-utterance setup?
 	l.ch <- l.utterance
+
+	// TODO: this will eat a lot of memory
 	l.ParsedRules = append(l.ParsedRules, l.utterance)
+
 	l.utterance = NewUtterance()
 }
 
 func (l *NluRuleListener) EnterText(c *parser.TextContext) {
-	text := c.GetText()
-
-	new_text := make(map[string]string)
-	new_text["text"] = text
+	newText := map[string]string{
+		"text": c.GetText(),
+	}
 
-	if l.utterance.current_entity != "" {
-		new_text["entity"] = l.utterance.current_entity
+	if l.utterance.CurrentEntity != "" {
+		newText["entity"] = l.utterance.CurrentEntity
 	}
-	if l.utterance.current_intent != "" {
-		new_text["intent"] = l.utterance.current_intent
-		new_text["intent_bio"] = l.utterance.current_intent_bio
-		if l.utterance.current_intent_bio == "b"{
-			l.utterance.current_intent_bio = "i"
+
+	if l.utterance.CurrentIntent != "" {
+		newText["intent"] = l.utterance.CurrentIntent
+		newText["intent_bio"] = l.utterance.CurrentIntentBio
+
+		if l.utterance.CurrentIntentBio == IntentBioB {
+			l.utterance.CurrentIntentBio = IntentBioI
 		}
 	}
 
-	l.utterance.Parsed = append(l.utterance.Parsed, new_text)
+	l.utterance.Parsed = append(l.utterance.Parsed, newText)
 }
 
 func (l *NluRuleListener) EnterIndent(c *parser.IndentContext) {
-	new_text := make(map[string]string)
-	new_text["text"] = " "
-	l.utterance.Parsed = append(l.utterance.Parsed, new_text)
+	l.utterance.Parsed = append(
+		l.utterance.Parsed,
+		map[string]string{
+			"text": TextEmptySpace,
+		},
+	)
 }
 
 func (l *NluRuleListener) EnterEntity(c *parser.EntityContext) {
-	entity_name, ok := c.Entity_name().(*parser.Entity_nameContext)
-	if !ok || entity_name == nil {
-		panic("failed to cast type") // TODO: graceful error handling
+	e := c.Entity_name()
+	entityName, ok := e.(*parser.Entity_nameContext)
+	if (!ok || entityName == nil) && l.debug {
+		fmt.Printf("failed to parse entity name: %+v\n", e)
```

Comment:
There should be a proper logger being passed to the listener, but for now I just hacked it around. Having panics in production is generally a bad idea 🙂